analysers/analyser_osmosis_indoor.py

```diff
@@ -0,0 +1,204 @@
+##                                                                       ##
+###########################################################################
+
+from modules.OsmoseTranslation import T_
+from .Analyser_Osmosis import Analyser_Osmosis
+
+sql00 = """
+CREATE TEMP TABLE indoor_surfaces AS
+SELECT
+    id,
+    linestring AS geom,
+    nodes,
+    tags->'indoor' AS indoor,
+    tags->'level' AS level,
+    tags,
+    (NOT tags?'access' OR NOT tags->'access' IN ('no', 'private')) AS public_access
+FROM
+    ways
+WHERE
+    is_polygon AND
+    tags != ''::hstore AND
+    tags?'indoor' AND
+    tags->'indoor' in ('room', 'corridor', 'area', 'level')
+"""
+
+sql01 = """
+CREATE INDEX indoor_surfaces_idx_geom on indoor_surfaces USING gist(geom)
+"""
+
+sql10 = """
+SELECT
+    id,
+    ST_AsText(way_locate(linestring))
+FROM
+    ways
+WHERE
+    tags != ''::hstore AND
+    NOT is_polygon AND
+    tags?'indoor' AND
+    tags->'indoor' in ('room', 'corridor', 'area', 'level')
+"""
+
+sql20 = """
+SELECT
+    id,
+    ST_AsText(geom)
+FROM
+    nodes
+WHERE
+    tags != ''::hstore AND
+    tags?'indoor' AND
+    tags->'indoor' in ('room', 'corridor', 'area', 'level')
+"""
+
+sql30 = """
+SELECT
+    public_indoor_rooms.id,
+    ST_AsText(way_locate(public_indoor_rooms.geom))
+FROM
+    indoor_surfaces AS public_indoor_rooms
+    LEFT JOIN nodes ON
+        public_indoor_rooms.geom && nodes.geom AND
+        nodes.id = ANY(public_indoor_rooms.nodes) AND
+        nodes.tags != ''::hstore AND
+        nodes.tags?'door'
+WHERE
+    public_indoor_rooms.indoor = 'room' AND
+    public_indoor_rooms.public_access AND
+    nodes.id IS NULL
+"""
+
+sql40 = """
+SELECT
+    id,
+    ST_AsText(way_locate(geom))
+FROM
+    indoor_surfaces
+WHERE
+    NOT tags?'level'
+"""
+
+sql50 = """
+SELECT
+    id,
+    ST_AsText(way_locate(geom))
+FROM
+    indoor_surfaces
+WHERE
+    indoor_surfaces.indoor != 'room' AND
+    indoor_surfaces.tags?'shop'
+"""
+
+sql60 = """
+CREATE TEMP TABLE indoor_surfaces_connected_to_highways AS
+SELECT DISTINCT
+    indoor_surfaces.id,
+    indoor_surfaces.geom,
+    indoor_surfaces.indoor,
+    indoor_surfaces.level AS surface_level,
+    highway_ends.nid,
+    highway_ends.highway,
+    highways.tags->'level' AS connected_highway_level
+FROM
+    indoor_surfaces
+    JOIN nodes ON
+        nodes.geom && indoor_surfaces.geom AND
+        nodes.id = ANY(indoor_surfaces.nodes)
+    JOIN highway_ends ON
+        highway_ends.nid = nodes.id
+    JOIN highways ON
+        highways.id = highway_ends.id
+WHERE
+    indoor_surfaces.indoor IN ('room', 'corridor', 'area') AND
+    highways.highway IN ('steps', 'footway', 'pedestrian')
+"""
+
+# TODO : check that all surfaces have at least one connected_highway_level that matches it own surface_level
+# (dealing with level="0;1" that occurs on highway=steps
+# and assuming than no level on highway is probably implicit level=0)
```

Comment:
Split the string as array and use array intersection `&&` ?

Reply:
yes, it would work for sure, but I'm still unsure about this one. I think it is not a high priority as the current version (without level checking) does not produce false positives.

I'm a bit afraid that this check will be very confusing.
For example, if you have an indoor area on level 1, connected by a staircase with no level tag, the actual issue is not that "the indoor surface is not connected to another feature or to a footpath", but that we need the level tag on the stairs in this case.

I don't have very clear ideas on the subject yet, I guess that ideally, we would manage to have tests that guide us a little more in the resolution.